### PR TITLE
Rework Fullscreen::Borderless enum variant

### PR DESCRIPTION
**Pull request by kchibisov**
This changes `Fullscreen::Borderless` enum variant from
`Fullscreen::Borderless(MonitorHandle)` to
`Fullscreen::Borderless(Option<MonitorHandle>)`. Providing
`None` to it will result in picking system's preferred monitor.

This is an implementation of proposal from https://github.com/rust-windowing/winit/issues/793#issuecomment-676820743 . I can only speak for Wayland implementation here. So `macOS`,`ios`,`X11`, and `Windows`, implementations are interested and likely wrong or inefficient to do what this variant should do.

cc for checking implementations
@francesca64 (not sure if you're active checking macOS/ios stuff now, let me know if you don't want to take a look)
@Osspial (windows) The implementation I've provided needs fixing for sure, since I have a hard time understand how windows even works in winit.
@murarth  (X11)

### CODE REVIEW

File: src/platform_impl/windows/event_loop.rs [hunk 1]
```diff
@@ -851,7 +851,8 @@ unsafe extern "system" fn public_window_callback<T: 'static>(
                     winuser::MonitorFromRect(&new_rect, winuser::MONITOR_DEFAULTTONULL);
                 match fullscreen {
                     Fullscreen::Borderless(ref mut fullscreen_monitor) => {
-                        if new_monitor != fullscreen_monitor.inner.hmonitor()
+                        // FIXME This should be properly implemented.
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply src/platform_impl/windows/event_loop.rs]: Done

**Comment by kchibisov** [comment]
I've implemented parts on all platforms and tested at least X11 and Wayland. The impl for macOS, ios, and windows are implemented in conservative way by mapping `None` to `current_monitor` early on.

File: CHANGELOG.md [hunk 2]
```diff
@@ -33,6 +33,7 @@
 - On macOS, updated core-* dependencies and cocoa.
 - Bump `parking_lot` to 0.11
 - On Android, bump `ndk`, `ndk-sys` and `ndk-glue` to 0.2. Checkout the new ndk-glue main proc attribute.
+- **Breaking:** `Fullscreen` enum now uses `Option<Borderless>` instead of `Borderless` to pick system preferred monitor.
```
Comment (chrisduerr) [inline CHANGELOG.md:36]: `Borderless(Option<MonitorHandle>)`, not `Option<Borderless>`, right?

File: examples/window_debug.rs [hunk 3]
```diff
@@ -85,10 +86,17 @@ fn main() {
                         if window.fullscreen().is_some() {
                             window.set_fullscreen(None);
                         } else {
-                            let monitor = window.current_monitor().unwrap();
+                            let monitor = window.current_monitor();
                             window.set_fullscreen(Some(Fullscreen::Borderless(monitor)));
```
Comment (chrisduerr) [inline examples/window_debug.rs:90]: Isn't this the same as Borderless(None)?

File: src/platform_impl/ios/view.rs [hunk 4]
```diff
@@ -519,7 +519,15 @@ pub unsafe fn create_window(
             msg_send![window, setScreen:video_mode.monitor().ui_screen()]
         }
         Some(Fullscreen::Borderless(ref monitor)) => {
-            msg_send![window, setScreen:monitor.ui_screen()]
+            let uiscreen: id = match monitor {
+                Some(ref monitor) => monitor.ui_screen() as id,
+                None => {
+                    let uiscreen: id = msg_send![window, screen];
+                    uiscreen
```
Comment (chrisduerr) [inline src/platform_impl/ios/view.rs:526]: Type inference issues I'd assume?

File: src/platform_impl/ios/view.rs [hunk 5]
```diff
@@ -519,7 +519,15 @@ pub unsafe fn create_window(
             msg_send![window, setScreen:video_mode.monitor().ui_screen()]
         }
         Some(Fullscreen::Borderless(ref monitor)) => {
-            msg_send![window, setScreen:monitor.ui_screen()]
+            let uiscreen: id = match monitor {
+                Some(ref monitor) => monitor.ui_screen() as id,
```
Comment (chrisduerr) [inline src/platform_impl/ios/view.rs:523]: Usually should use `match &monitor` instead of `Some(ref monitor)`. There's a surprising amount of `ref`s here, given that they're almost never necessary in Rust anymore.

File: src/platform_impl/ios/window.rs [hunk 6]
```diff
@@ -196,7 +196,10 @@ impl Inner {
                     let () = msg_send![uiscreen, setCurrentMode: video_mode.video_mode.screen_mode];
                     uiscreen
                 }
-                Some(Fullscreen::Borderless(monitor)) => monitor.ui_screen() as id,
+                Some(Fullscreen::Borderless(Some(monitor))) => monitor.ui_screen() as id,
+                Some(Fullscreen::Borderless(None)) => {
+                    self.current_monitor_inner().ui_screen() as id
+                }
```
Comment (chrisduerr) [inline src/platform_impl/ios/window.rs:202]: Could in theory combine these and do something like `monitor.unwrap_or_else(|_| self.current_monitor_inner()).ui_screen() as id` maybe?

File: src/platform_impl/ios/window.rs [hunk 7]
```diff
@@ -353,8 +356,10 @@ impl Window {
                 Some(Fullscreen::Exclusive(ref video_mode)) => {
                     video_mode.video_mode.monitor.ui_screen() as id
                 }
-                Some(Fullscreen::Borderless(ref monitor)) => monitor.ui_screen() as id,
-                None => monitor::main_uiscreen().ui_screen(),
+                Some(Fullscreen::Borderless(Some(ref monitor))) => monitor.inner.ui_screen(),
+                Some(Fullscreen::Borderless(None)) | None => {
+                    monitor::main_uiscreen().ui_screen() as id
+                }
```
Comment (chrisduerr) [inline src/platform_impl/ios/window.rs:362]: I'd assume this is because we don't have any monitor yet? Should this be documented maybe that it's not something supported at startup?

File: src/platform_impl/linux/wayland/window.rs [hunk 8]
```diff
@@ -154,11 +154,18 @@ impl Window {
...
-            Some(Fullscreen::Borderless(_)) => unreachable!(),
+            Some(Fullscreen::Borderless(monitor)) => {
+                let monitor = if let Some(RootMonitorHandle {
+                    inner: PlatformMonitorHandle::Wayland(monitor),
+                }) = monitor
+                {
+                    Some(monitor.proxy)
+                } else {
+                    None
+                };
```
Comment (chrisduerr) [inline src/platform_impl/linux/wayland/window.rs:165]: This looks suspiciously `map`y.

File: src/platform_impl/linux/wayland/window.rs [hunk 9]
```diff
@@ -348,16 +357,18 @@ impl Window {
...
-                    .set_fullscreen(Some(&monitor_id.proxy));
+            Some(Fullscreen::Borderless(monitor)) => {
+                let monitor = if let Some(RootMonitorHandle {
+                    inner: PlatformMonitorHandle::Wayland(monitor),
+                }) = monitor
+                {
+                    Some(monitor.proxy)
+                } else {
+                    None
+                };
```
Comment (chrisduerr) [inline src/platform_impl/linux/wayland/window.rs:368]: This also looks very `map`y.

File: src/platform_impl/linux/x11/window.rs [hunk 10]
```diff
@@ -585,6 +585,16 @@ impl UnownedWindow {
 
     fn set_fullscreen_inner(&self, fullscreen: Option<Fullscreen>) -> Option<util::Flusher<'_>> {
         let mut shared_state_lock = self.shared_state.lock();
+        // Map Borderless(`None`) to current monitor.
```
Comment (chrisduerr) [inline src/platform_impl/linux/x11/window.rs:588]:
```suggestion
        let mut shared_state_lock = self.shared_state.lock();
        
        // Map Borderless(`None`) to current monitor.
```

I'd just give it a touch of space.

File: src/platform_impl/linux/x11/window.rs [hunk 11]
```diff
@@ -585,6 +585,16 @@ impl UnownedWindow {
...
+        let fullscreen = match fullscreen {
+            Some(Fullscreen::Borderless(None)) => {
+                let current_monitor = Some(RootMonitorHandle {
+                    inner: PlatformMonitorHandle::X(self.current_monitor()),
+                });
+                Some(Fullscreen::Borderless(current_monitor))
+            }
+            _ => fullscreen,
+        };
```
Comment (chrisduerr) [inline src/platform_impl/linux/x11/window.rs:597]: Could also make it a `mut fullscreen` in parameters and use `if let Some` here, right? I think that might come out a couple lines shorter.

File: src/platform_impl/linux/x11/window.rs [hunk 12]
```diff
@@ -649,10 +659,12 @@ impl UnownedWindow {
...
-                    }) => (Some(video_mode), video_mode.monitor.as_ref().unwrap()),
-                    Fullscreen::Borderless(RootMonitorHandle {
-                        inner: PlatformMonitorHandle::X(ref monitor),
-                    }) => (None, monitor),
+                    }) => (Some(video_mode), video_mode.monitor.clone().unwrap()),
+                    Fullscreen::Borderless(Some(RootMonitorHandle {
+                        inner: PlatformMonitorHandle::X(monitor),
+                    })) => (None, monitor),
+                    // We've mapped `None` in borderless earlier.
+                    Fullscreen::Borderless(None) => unreachable!(),
```
Comment (chrisduerr) [inline src/platform_impl/linux/x11/window.rs:667]: This looks a bit unfortunate to me. Why does it need to be remapped earlier instead of handled here? So comparison to the existing state works? I don't really see where the monitor is needed in the old/new comparison.

File: src/platform_impl/windows/event_loop.rs [hunk 13]
```diff
@@ -853,7 +853,9 @@ unsafe extern "system" fn public_window_callback<T: 'static>(
                     winuser::MonitorFromRect(&new_rect, winuser::MONITOR_DEFAULTTONULL);
                 match fullscreen {
                     Fullscreen::Borderless(ref mut fullscreen_monitor) => {
-                        if new_monitor != fullscreen_monitor.inner.hmonitor()
+                        // Unwrap is safe, since we explicitly map `None` to current monitor
+                        // when we're setting fullscreen state.
+                        if new_monitor != fullscreen_monitor.as_ref().unwrap().inner.hmonitor()
```
Comment (chrisduerr) [inline src/platform_impl/windows/event_loop.rs:858]: This just seems ripe for failure. I don't like the mapping things and then making assumptions later structure.

File: src/window.rs [hunk 14]
```diff
@@ -624,6 +624,10 @@ impl Window {
 
     /// Gets the window's current fullscreen state.
     ///
+    /// If `None` in `Fullscreen::Borderless` was returned it means
+    /// that the window is in a `Fullscreen` state, however there's
+    /// no monitor information available.
```
Comment (chrisduerr) [inline src/window.rs:629]: This seems a bit problematic, because we seem to conflate the concepts of "no monitor information available" and "current / primary monitor". Like a user doesn't request fullscreen on no monitor, that doesn't make sense.

File: src/window.rs [hunk 15]
```diff
@@ -850,10 +854,15 @@ impl Default for CursorIcon {
     }
 }
 
+/// Fullscreen mode to be set on certain output.
+///
+/// Providing `None` to `Borderless` will result in fullscreen
+/// being set on a system preferred monitor. In general it'll be
+/// the current monitor.
```
Comment (chrisduerr) [inline src/window.rs:861]: I'm not sure if I would say something about the "system preferred monitor" here, it could be easily confused with the primary monitor. Which is never used here afaics?

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (kchibisov) [reply CHANGELOG.md:36]: indeed.

File: examples/window_debug.rs (hunk 3 above)
Comment (kchibisov) [reply examples/window_debug.rs:90]: Well, yeah, but that's purely for testing. Could be different on Wayland in certain cases though.

File: src/platform_impl/ios/view.rs (hunk 4 above)
Comment (kchibisov) [reply src/platform_impl/ios/view.rs:526]: Yeah, it's really strange how it worked, but returning `msg_send!` wasn't working. I'm not sure that there's a CI log for that, but it was somewhere.

File: src/platform_impl/ios/view.rs (hunk 5 above)
Comment (kchibisov) [reply src/platform_impl/ios/view.rs:523]: that's a pretty old rust iirc, so that's why it's using ref, I've just mirrored existing code.

File: src/platform_impl/ios/window.rs (hunk 7 above)
Comment (kchibisov) [reply src/platform_impl/ios/window.rs:362]: It's basically that we're asking for a current monitor. However if you look at ios it doesn't really have a fullscreen in a way you'd know. It just always sets screen and it's always fullscreened, that branch just picks the 'preffered monitor', which is the current one here. Not sure whether such thing should be documented, since it's there for a long time.

File: src/window.rs (hunk 14 above)
Comment (kchibisov) [reply src/window.rs:629]: I mean what should I describe here then? It can clearly be returned with `None` on GNOME(Wayland) :P

File: src/window.rs (hunk 15 above)
Comment (kchibisov) [reply src/window.rs:861]: Yeah, we can reword it to be current monitor?

File: src/platform_impl/windows/event_loop.rs (hunk 13 above)
Comment (kchibisov) [reply src/platform_impl/windows/event_loop.rs:858]: Well, you can't store it with `None`, so you always store it unwraped, one more thing that you can't access monitor here.

File: src/platform_impl/linux/wayland/window.rs (hunk 8 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/window.rs:165]: The reason why there's no `map` its' because there's an `X` variant on `PlatformMonitorHandle` so I need `if let` anyway. And that `X` variant is presented based on feature, so in one case you must put `unreachable!` there, so I decided to you that, since shorter.

File: src/platform_impl/linux/wayland/window.rs (hunk 9 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/window.rs:368]: The reason why there's no `map` its' because there's an `X` variant on `PlatformMonitorHandle` so I need `if let` anyway. And that `X` variant is presented based on feature, so in one case you must put `unreachable!` there, so I decided to you that, since shorter.

File: src/platform_impl/linux/x11/window.rs (hunk 11 above)
Comment (kchibisov) [reply src/platform_impl/linux/x11/window.rs:597]: I'll just match it later on instead of expanding early on

File: src/platform_impl/windows/event_loop.rs (hunk 13 above)
Comment (kchibisov) [reply src/platform_impl/windows/event_loop.rs:858]: I'll ask for current monitor, since it should be available here.

Comment (kchibisov) [reply src/platform_impl/windows/event_loop.rs]: Oh, nvm, it's acually important to store those things unwrapped and not in Option for fullscreen. So how we should approach that? I'm fine with `unwrap` here tbqh.

File: src/platform_impl/ios/window.rs (hunk 6 above)
Comment (kchibisov) [reply src/platform_impl/ios/window.rs]: Yeah, it should work.

File: src/platform_impl/windows/window.rs [hunk 16]
```diff
@@ -393,6 +393,17 @@ impl Window {
...
+        let fullscreen = match fullscreen {
+            Some(Fullscreen::Borderless(None)) => {
+                let current_monitor = Some(RootMonitorHandle {
+                    inner: monitor::current_monitor(window.0),
+                });
+
+                Some(Fullscreen::Borderless(current_monitor))
+            }
+            _ => fullscreen,
+        };
```
Comment (chrisduerr) [inline src/platform_impl/windows/window.rs:405]: Why can't this be stored with None as Borderless? Do you want to make it possible to go from fullscreen on one monitor to fullscreen on another using None as parameter? When would you ever want to do that?

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: src/platform_impl/windows/window.rs (hunk 16 above)
Comment (kchibisov) [reply src/platform_impl/windows/window.rs:405]: >Why can't this be stored with None as Borderless? Do you want to make it possible to go from fullscreen on one monitor to fullscreen on another using None as parameter? When would you ever want to do that?

it's due to monitor logic in event_loop.rs when you can move fullscreren window across the monitor, since I bet the current monitor after move is a new monitor, so you won't be able to update things properly. We may always update monitor there if the current one is `None` though.

File: src/platform_impl/linux/x11/window.rs [hunk 17]
```diff
@@ -649,10 +649,11 @@ impl UnownedWindow {
...
                         video_mode: PlatformVideoMode::X(ref video_mode),
-                    }) => (Some(video_mode), video_mode.monitor.as_ref().unwrap()),
-                    Fullscreen::Borderless(RootMonitorHandle {
-                        inner: PlatformMonitorHandle::X(ref monitor),
-                    }) => (None, monitor),
+                    }) => (Some(video_mode), video_mode.monitor.clone().unwrap()),
+                    Fullscreen::Borderless(Some(RootMonitorHandle {
+                        inner: PlatformMonitorHandle::X(monitor),
+                    })) => (None, monitor),
+                    Fullscreen::Borderless(None) => (None, self.current_monitor()),
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply src/platform_impl/linux/x11/window.rs:656]: >I'd assume you changed this to a owned instead of a reference because Rust doesn't allow you to return &self.current_monitor() due to lifetime guarantees here?

Yes.

**Comment by kchibisov** [comment]
@chrisduerr I've updated PR, let me know if it's fine now.

File: src/platform_impl/windows/event_loop.rs [hunk 18]
```diff
@@ -853,8 +853,14 @@ unsafe extern "system" fn public_window_callback<T: 'static>(
...
-                            && new_monitor != ptr::null_mut()
+                        if new_monitor == ptr::null_mut() {
+                            return 0;
+                        }
+
+                        if fullscreen_monitor
+                            .as_ref()
+                            .map(|monitor| new_monitor != monitor.inner.hmonitor())
+                            .unwrap_or(true)
                         {
```
Comment (chrisduerr) [inline src/platform_impl/windows/event_loop.rs:864]: Any reason why these aren't combined anymore? Seems strange to have one thing as a gate in the front and the other that does the same thing as a dedicated block.

File: src/window.rs [hunk 19]
```diff
@@ -628,6 +628,8 @@ impl Window {
     ///
     /// - **iOS:** Can only be called on the main thread.
     /// - **Android:** Will always return `None`.
+    /// - **Wayland** Can return `Borderless(None)` in some scenarios
+    ///
```
Comment (chrisduerr) [inline src/window.rs:632]:
```suggestion
    /// - **Wayland** Can return `Borderless(None)` in some scenarios.
    ///
```

I personally am not a huge fan of the "in some scenarios". Can we be more concrete here?

File: src/window.rs [hunk 20]
```diff
@@ -850,10 +852,14 @@ impl Default for CursorIcon {
     }
 }
 
+/// Fullscreen mode to be set on certain output.
```
Comment (chrisduerr) [inline src/window.rs:855]:
```suggestion
/// Fullscreen modes.
```

The certain output stuff isn't really true since it might not be a specific output and Exclusive takes a VideoMode instead?

File: src/window.rs [hunk 21]
```diff
@@ -850,10 +852,14 @@ impl Default for CursorIcon {
...
+///
+/// Providing `None` to `Borderless` will result in fullscreen
+/// being set on a current monitor.
```
Comment (chrisduerr) [inline src/window.rs:858]: Maybe this is better as a doc comment on the `Borderless` enum variant?

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: src/platform_impl/linux/wayland/window.rs [hunk 22]
```diff
@@ -154,11 +154,17 @@ impl Window {
...
-            #[cfg(feature = "x11")]
-            Some(Fullscreen::Borderless(_)) => unreachable!(),
+            Some(Fullscreen::Borderless(monitor)) => {
+                let monitor = monitor
+                    .map(|RootMonitorHandle { inner: monitor }| match monitor {
+                        PlatformMonitorHandle::Wayland(monitor) => Some(monitor.proxy),
+                        #[cfg(feature = "x11")]
+                        PlatformMonitorHandle::X(_) => None,
+                    })
+                    .flatten();
```
Comment (chrisduerr) [inline src/platform_impl/linux/wayland/window.rs:164]:
```suggestion
                let monitor = monitor
                    .and_then(|RootMonitorHandle { inner: monitor }| match monitor {
                        PlatformMonitorHandle::Wayland(monitor) => Some(monitor.proxy),
                        #[cfg(feature = "x11")]
                        PlatformMonitorHandle::X(_) => None,
                    });
```

File: src/platform_impl/linux/wayland/window.rs [hunk 23]
```diff
@@ -348,16 +356,17 @@ impl Window {
...
-                    .unwrap()
-                    .set_fullscreen(Some(&monitor_id.proxy));
+            Some(Fullscreen::Borderless(monitor)) => {
+                let monitor = monitor
+                    .map(|RootMonitorHandle { inner: monitor }| match monitor {
+                        PlatformMonitorHandle::Wayland(monitor) => Some(monitor.proxy),
+                        #[cfg(feature = "x11")]
+                        PlatformMonitorHandle::X(_) => None,
+                    })
+                    .flatten();
```
Comment (chrisduerr) [inline src/platform_impl/linux/wayland/window.rs:366]:
```suggestion
                let monitor = monitor
                    .and_then(|RootMonitorHandle { inner: monitor }| match monitor {
                        PlatformMonitorHandle::Wayland(monitor) => Some(monitor.proxy),
                        #[cfg(feature = "x11")]
                        PlatformMonitorHandle::X(_) => None,
                    });
```

File: src/window.rs [hunk 24]
```diff
@@ -628,6 +628,8 @@ impl Window {
     ///
     /// - **iOS:** Can only be called on the main thread.
     /// - **Android:** Will always return `None`.
+    /// - **Wayland:** Can return `Borderless(None)` when no monitors are presented.
```
Comment (chrisduerr) [inline src/window.rs:631]: What do you mean by "no monitors are presented"? Like no monitors are *present*? Usually a monitor does the presenting and isn't presented, right?

File: src/window.rs [hunk 25]
```diff
@@ -628,6 +628,8 @@ impl Window {
...
+    ///
```
Comment (chrisduerr) [inline src/window.rs:632]:
```suggestion
```

File: src/window.rs [hunk 26]
```diff
@@ -850,10 +852,14 @@ impl Default for CursorIcon {
...
 }
 
+/// Fullscreen modes.
 #[derive(Clone, Debug, PartialEq)]
 pub enum Fullscreen {
     Exclusive(VideoMode),
-    Borderless(MonitorHandle),
+
+    /// Providing `None` to `Borderless` will result in fullscreen
+    /// being set on a current monitor.
```
Comment (chrisduerr) [inline src/window.rs:861]:
```suggestion
    /// Providing `None` to `Borderless` will fullscreen on the current monitor.
```

I think this is a bit better. Since "fullscreen is set on..." doesn't really sound right. And either way it should be "**the** current monitor".

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: src/window.rs (hunk 24 above)
Comment (kchibisov) [reply src/window.rs:631]: It means that user powered off every monitor it had, however our window was fullscreened when the user done that, so it's fullscreened, but there are no outputs.

**Review by chrisduerr — APPROVED** [review]
LGTM
